### ISSUE TITLE
hxcpp 3.2 regression: re-assignment of a dynamic variable to a different type silently does nothing

### ISSUE THREAD

**Issue by bjitivo** [HaxeFoundation/hxcpp#252]
Here is a sample code:

```
class Foo
{
    public static function main()
    {
        var json = haxe.Json.parse("{\"array\":[\"Hello\"]}");

        var field = Reflect.field(json, "array");
        
        if (Std.is(field, Array)) {
            var array : Array<Dynamic> = field;
            if (0 < array.length) {
                field = array[0];
            }
        }

        trace("field is " + field);
    }
}
```

The variable 'field' is re-used both as an Array, and as a value within the Array.  But since the type of 'field' is Dynamic (as inferred by the return type of Reflect.field()), assigning it to any value subsequently should work.

Expected result:

<pre>
field is Hello
</pre>

Actual results:

- haxe 3.1.3, neko target: as expected
- haxe 3.1.3, hxcpp target: as expected
- haxe 3.2, neko target: as expected
- haxe 3.2, hxcpp target: invalid result:
<pre>
Foo.hx:17: field is null
</pre>

**Comment by Simn** [comment]
I just tested this and it's still failing the same way.

**Comment by hughsando** [comment]
@Simn I think this is a unification error.
```
class Test
{
    public static function main()
    {
        var json = haxe.Json.parse("{\"array\":[\"Hello\"]}");
        var field = Reflect.field(json, "array");

        var array : Array<Dynamic> = field;
        field = field[0];

        trace("field is " + field);
    }
}
```
From the dump, 'field' is being typed as "```Array<Dynamic>```", even though it takes on value of field[0], which should be 'Dynamic', but I think it is getting inferred as "```Array<Dynamic>```" because of the assign:
```
			[Binop:Array<Dynamic>]
				[Local field(337):Array<Dynamic>]
				=
				[Array:Dynamic]
					[Local field(337):Array<Dynamic>]
					[Const:Int] 0
```
perhaps due to some circular logic.

**Comment by Simn** [comment]
This happens because Dynamic is never bound to a monomorph because Nicolas doesn't want to "spread" it like a disease. While I sympathize with the notion, it leads to errors in several cases such as this one. The unification happens on the line `var array : Array<Dynamic> = field;` which binds `field`'s monomorph to `Array<Dynamic>`.

Unfortunately, that's how it is at the moment and I'm not sure it's going to change.

**Comment by hughsando** [comment]
Yes, I thought so.  It is really only a regression because hxcpp 3.1 did a weaker job of respecting the ```Array<Dynamic>``` type.  The best I can do I think would be to throw a runtime error in this case.

**Comment by Simn** [comment]
I think that's better than silently failing.

I wish we could blame the user for this, but understanding the behavior requires a bit of internal knowledge how monomorphs and Dynamic are treated. It's quite unfortunate, but I can't really do anything about it either at the moment...
